src/plugins/mavlink_ftp/mavlink_ftp_impl.cpp

```diff
@@ -0,0 +1,1121 @@
+    std::lock_guard<std::mutex> lock(_curr_op_mutex);
+    if (_curr_op != CMD_NONE) {
+        callback(MavlinkFTP::Result::IN_PROGRESS);
+        return;
+    }
+    if (from_path.length() + to_path.length() + 1 >= max_data_length) {
+        callback(MavlinkFTP::Result::INVALID_PARAMETER);
+        return;
+    }
+
+    uint8_t raw_payload[MAVLINK_MSG_FILE_TRANSFER_PROTOCOL_FIELD_PAYLOAD_LEN];
+    PayloadHeader* payload = reinterpret_cast<PayloadHeader*>(raw_payload);
+    payload->seq_number = _seq_number++;
+    payload->session = 0;
+    payload->opcode = _curr_op = CMD_RENAME;
+    payload->offset = 0;
+    strncpy(reinterpret_cast<char*>(&payload->data[0]), from_path.c_str(), max_data_length);
+    payload->size = from_path.length() + 1;
+    strncpy(reinterpret_cast<char*>(&payload->data[payload->size]), to_path.c_str(), max_data_length - payload->size);
+    payload->size += to_path.length() + 1;
+    _curr_op_result_callback = callback;
+    _send_mavlink_ftp_message(raw_payload);
+}
+
+void MavlinkFTPImpl::calc_file_crc32_async(
+    const std::string& path, MavlinkFTP::file_crc32_result_callback_t callback)
+{
+    std::lock_guard<std::mutex> lock(_curr_op_mutex);
+    if (_curr_op != CMD_NONE) {
+        callback(MavlinkFTP::Result::IN_PROGRESS, 0);
+        return;
+    }
+    if (path.length() >= max_data_length) {
+        callback(MavlinkFTP::Result::INVALID_PARAMETER, 0);
+        return;
+    }
+
+    uint8_t raw_payload[MAVLINK_MSG_FILE_TRANSFER_PROTOCOL_FIELD_PAYLOAD_LEN];
+    PayloadHeader* payload = reinterpret_cast<PayloadHeader*>(raw_payload);
+    payload->seq_number = _seq_number++;
+    payload->session = 0;
+    payload->opcode = _curr_op = CMD_CALC_FILE_CRC32;
+    payload->offset = 0;
+    strncpy(reinterpret_cast<char*>(payload->data), path.c_str(), max_data_length);
+    payload->size = path.length() + 1;
+    _current_crc32_result_callback = callback;
+    _send_mavlink_ftp_message(raw_payload);
+}
+
+void MavlinkFTPImpl::_send_mavlink_ftp_message(uint8_t* raw_payload)
+{
+    // LogInfo() << "_send_mavlink_ftp_message";
+    std::lock_guard<std::mutex> lock(_timer_mutex);
+    mavlink_msg_file_transfer_protocol_pack(
+        _parent->get_own_system_id(),
+        _parent->get_own_component_id(),
+        &_last_command,
+        _network_id,
+        _parent->get_system_id(),
+        _get_target_component_id(),
+        raw_payload);
+    _parent->send_message(_last_command);
+
+    _reset_timer();
+    if (!_last_command_timer_running) {
+        _last_command_timer_running = true;
+        _last_command_timer = std::async(std::launch::async, [&]() {
+            while (true) {
+                _timer_mutex.lock();
+                if (!_last_command_timer_running) {
+                    _timer_mutex.unlock();
+                    return;
+                }
+                auto now = std::chrono::steady_clock::now();
+                auto time_diff = std::chrono::duration_cast<std::chrono::milliseconds>(
```

Comment:
I'd recommend to use the included timer if possible. Have a look at:
https://github.com/mavlink/MAVSDK/blob/f4065b3fb56e65674d045a3406cfb9c2536fcf65/src/core/system_impl.h#L56-L58

Reply:
Changed to use included timer.